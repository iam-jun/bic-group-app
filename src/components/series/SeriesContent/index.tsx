import React, { FC, useState } from 'react';
import { View, StyleSheet } from 'react-native';
import { ExtendedTheme, useTheme } from '@react-navigation/native';
import Text from '~/baseComponents/Text';

import DescriptionSection from './DescriptionSection';
import TitleSection from './TitleSection';
import ListItem from './ListItem';

import { spacing } from '~/theme';
import { Button } from '~/baseComponents';
import { useRootNavigation } from '~/hooks/navigation';
import seriesStack from '~/router/navigator/MainStack/stacks/series/stack';
import { IPost, PostType } from '~/interfaces/IPost';
import { trackEvent } from '~/services/tracking';
import { TrackingEventContentReadProperties } from '~/services/tracking/Interface';
import { TrackingEventContentReadAction, TrackingEvent } from '~/services/tracking/constants';

const LIMIT_ITEM = 3;

type SeriesContentProps = {
  series: IPost;
  isLite?: boolean;
};

const SeriesContent: FC<SeriesContentProps> = ({ series, isLite }) => {
  const theme: ExtendedTheme = useTheme();
  const { colors } = theme;

  const {
    id,
    title,
    updatedAt,
    coverMedia,
    summary,
    items,
    titleHighlight,
    summaryHighlight,
  } = series || {};
  const { rootNavigation } = useRootNavigation();
  const [itemShowAll, setItemShowAll] = useState(false);
  const listItem = itemShowAll ? items : items?.slice?.(0, LIMIT_ITEM);
  const titleSection = isLite && titleHighlight ? titleHighlight : title;
  const summarySection = isLite && summaryHighlight ? summaryHighlight : summary;

  const goToSeriesDetail = () => {
    rootNavigation.navigate(seriesStack.seriesDetail, { seriesId: id });

    // tracking event
    const eventContentReadProperties: TrackingEventContentReadProperties = {
      content_type: PostType.SERIES,
      action: TrackingEventContentReadAction.BODY,
    };
    trackEvent({ event: TrackingEvent.CONTENT_READ, properties: eventContentReadProperties });
  };

  const onToggleShowItem = () => {
    setItemShowAll(!itemShowAll);
    // tracking event
    if (itemShowAll) {
      const eventContentReadProperties: TrackingEventContentReadProperties = {
        content_type: PostType.SERIES,
        action: TrackingEventContentReadAction.SEE_MORE,
      };
      trackEvent({ event: TrackingEvent.CONTENT_READ, properties: eventContentReadProperties });
    }
  };

  const renderListItem = () => (
    <ListItem
      listItem={listItem}
    />
  );

  const renderOptions = () => (
    <View style={[styles.default, !!isLite && styles.viewLite]}>
      {items?.length > LIMIT_ITEM && (
        <Text.SubtitleM
          testID="series_content.short_content"
          onPress={onToggleShowItem}
          color={colors.neutral60}
          useI18n
        >
          {itemShowAll
            ? 'common:text_see_less'
            : 'common:text_see_more'}
        </Text.SubtitleM>
      )}
    </View>
  );

  return (
    <View style={styles.container}>
      <Button onPress={goToSeriesDetail}>
        <TitleSection
          title={titleSection}
          time={updatedAt}
          numberOfItems={items?.length || 0}
          img={coverMedia?.url}
        />
      </Button>
      <DescriptionSection description={summarySection} />
      {renderListItem()}
      {renderOptions()}
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    paddingTop: spacing.padding.small,
    paddingHorizontal: spacing.padding.large,
    paddingBottom: spacing.padding.xSmall,
  },
  default: {
    paddingVertical: spacing.padding.base,
    marginTop: spacing.margin.base,
  },
  interestedUserCount: {
    paddingTop: 0,
    paddingBottom: 0,
    paddingHorizontal: 0,
    margin: 0,
    alignItems: 'flex-end',
  },
  viewLite: {
    paddingVertical: 0,
  },
});

export default SeriesContent;
