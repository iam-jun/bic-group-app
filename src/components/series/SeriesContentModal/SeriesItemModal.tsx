import React from 'react';
import { View, StyleSheet } from 'react-native';
import { ExtendedTheme, useTheme } from '@react-navigation/native';
import Text from '~/baseComponents/Text';
import Avatar from '~/baseComponents/Avatar';
import Button from '~/beinComponents/Button';
import { spacing } from '~/theme';
import ViewSpacing from '~/beinComponents/ViewSpacing';

interface ISeriesItemModalProps {
    data: any;
    onPressItem: () => void;
}

const SeriesItemModal: React.FC<ISeriesItemModalProps> = ({
  data,
  onPressItem,
}) => {
  const theme: ExtendedTheme = useTheme();
  const { colors } = theme;

  return (
    <Button
      style={styles.container}
      onPress={onPressItem}
      testID="series_item_modal.button"
    >
      <Avatar
        variant="small"
        source={data?.coverMedia?.url}
      />
      <ViewSpacing width={spacing.margin.small} />
      <View style={styles.boxTitle}>
        <Text.BodyM
          numberOfLines={1}
          color={colors.neutral60}
        >
          { data?.title }
        </Text.BodyM>
      </View>
    </Button>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    flexDirection: 'row',
    alignItems: 'center',
    paddingHorizontal: spacing.padding.large,
    paddingVertical: spacing.padding.small,
  },
  boxTitle: {
    flex: 1,
  },
});

export default SeriesItemModal;
