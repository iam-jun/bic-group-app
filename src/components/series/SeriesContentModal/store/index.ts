import { createStore, resetStore } from '~/store/utils';
import IBaseState, { InitStateType } from '~/store/interfaces/IBaseState';
import { IPost } from '~/interfaces/IPost';
import getSeriesByItems from './action/getSeriesByItems';
import { IGetSeries } from '~/interfaces/ISeries';

export interface ISeriesContentModalState extends IBaseState {
    series: {
        data: IPost[];
        loading: boolean;
        hasNextPage: boolean;
    };

    actions: {
        getSeriesByItems: (payload: IGetSeries) => void;
    };
}

const initState: InitStateType<ISeriesContentModalState> = {
  series: {
    data: [],
    loading: false,
    hasNextPage: true,
  },
};

const seriesContentModalStore = (set, get) => ({
  ...initState,

  actions: {
    getSeriesByItems: getSeriesByItems(set, get),
  },

  reset: () => resetStore(initState, set),
});

const useSeriesContentModalStore = createStore<ISeriesContentModalState>(seriesContentModalStore);

export default useSeriesContentModalStore;
