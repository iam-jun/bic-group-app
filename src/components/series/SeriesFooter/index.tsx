import React, { FC } from 'react';
import { View, StyleSheet } from 'react-native';

import ReactionView from '~/beinComponents/ReactionView';
import ContentFooter from '~/components/ContentView/ContentFooter';
import { IPost } from '~/interfaces/IPost';
import useContentActions from '~/hooks/useContentActions';
import { useBaseHook } from '~/hooks';
import { padding } from '~/theme/spacing';

interface SeriesFooterProps {
    series: IPost;
}

const SeriesFooter: FC<SeriesFooterProps> = ({ series }) => {
  const { t } = useBaseHook();

  const {
    id,
    ownerReactions,
    reactionsCount,
    commentsCount,
    setting,
  } = series || {};

  const {
    onAddReaction,
    onRemoveReaction,
    onLongPressReaction,
  } = useContentActions({ postId: id, ownerReactions, reactionsCount });

  const labelButtonComment = `${commentsCount ? `${commentsCount} ` : ''}${t(
    'post:button_comment',
  )}`;

  return (
    <View>
      <ReactionView
        style={styles.reactions}
        ownerReactions={ownerReactions}
        reactionsCount={reactionsCount}
        onAddReaction={onAddReaction}
        onRemoveReaction={onRemoveReaction}
        onLongPressReaction={onLongPressReaction}
      />
      <ContentFooter
        canReact={setting?.canReact}
        canComment={setting?.canComment}
        reactionsCount={reactionsCount}
        onAddReaction={onAddReaction}
        labelButtonComment={labelButtonComment}
      />
    </View>
  );
};

const styles = StyleSheet.create({
  reactions: {
    paddingHorizontal: padding.base,
  },
});

export default SeriesFooter;
