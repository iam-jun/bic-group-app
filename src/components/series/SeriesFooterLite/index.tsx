import React, { FC } from 'react';
import { ContentFooterLite } from '~/components/ContentView';
import { getTotalReactions } from '~/helpers/post';
import { IPost } from '~/interfaces/IPost';
import { formatLargeNumber } from '~/utils/formatData';

type SeriesFooterLiteProps = {
    series: IPost;
}

const SeriesFooterLite: FC<SeriesFooterLiteProps> = ({ series }) => {
  const {
    id, reactionsCount, commentsCount, totalUsersSeen,
  } = series;

  const numberOfReactions = formatLargeNumber(
    getTotalReactions(reactionsCount, 'user'),
  );

  return (
    <ContentFooterLite
      id={id}
      reactionsCount={Number(numberOfReactions)}
      commentsCount={commentsCount}
      totalUsersSeen={totalUsersSeen}
    />
  );
};

export default SeriesFooterLite;
