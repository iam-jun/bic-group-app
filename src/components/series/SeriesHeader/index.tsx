/* eslint-disable @typescript-eslint/no-unused-vars */
import React, { FC } from 'react';
import { Keyboard } from 'react-native';
import { Button } from '~/baseComponents';
import { ContentHeader } from '~/components/ContentView';
import { useBaseHook } from '~/hooks';
import { useUserIdAuth } from '~/hooks/auth';
import { useRootNavigation } from '~/hooks/navigation';
import { IAudienceGroup, IPost, PostType } from '~/interfaces/IPost';
import seriesStack from '~/router/navigator/MainStack/stacks/series/stack';
import { AlertDeleteAudiences } from '~/components/posts';
import useSeriesStore, { ISeriesState } from '~/screens/series/store';
import { PermissionKey } from '~/constants/permissionScheme';
import useMyPermissionsStore from '~/store/permissions';
import useModalStore from '~/store/modal';
import MenuContent from '~/components/MenuContent';

type SeriesHeaderProps = {
  series: IPost;
  disabled?: boolean;
};

const SeriesHeader: FC<SeriesHeaderProps> = ({ series, disabled }) => {
  const { actor, audience, id } = series;
  const { rootNavigation } = useRootNavigation();
  const userId = useUserIdAuth();
  const { t } = useBaseHook();

  const { showAlert, showModal } = useModalStore((state) => state.actions);

  const actions = useSeriesStore((state: ISeriesState) => state.actions);

  const { shouldHavePermissionOnSomeAudience } = useMyPermissionsStore((state) => state.actions);
  const canDeleteOwnPost = shouldHavePermissionOnSomeAudience(
    audience?.groups,
    PermissionKey.CRUD_POST_ARTICLE,
  );
  const isActor = actor?.id == userId;

  const goToSeriesDetail = () => {
    rootNavigation.navigate(seriesStack.seriesDetail, { seriesId: id });
  };

  const onPressHeader = () => {
    goToSeriesDetail();
  };

  const handleError = (listIdAudiences: string[]) => {
    if (listIdAudiences?.length <= 0 || audience?.groups?.length <= 0) {
      return;
    }

    const listAudiences = listIdAudiences.map((audienceId) => {
      const _audience = audience.groups.find(
        (audience: IAudienceGroup) => audience?.id === audienceId,
      );
      return _audience;
    });
    if (canDeleteOwnPost) {
      showAlert({
        title: t('series:title_delete_audiences_of_series'),
        children: (
          <AlertDeleteAudiences
            data={listAudiences}
            textContent={t('series:content_delete_audiences_of_series')}
          />
        ),
        cancelBtn: true,
        confirmLabel: t('common:text_remove'),
        ConfirmBtnComponent: Button.Danger,
        onConfirm: () => actions.removeAudiences(id, listIdAudiences),
        confirmBtnProps: { type: 'ghost' },
      });
    } else {
      showAlert({
        title: t('series:title_delete_audiences_of_series'),
        children: (
          <AlertDeleteAudiences
            data={listAudiences}
            textContent={t('series:content_not_able_delete_of_series')}
          />
        ),
        cancelBtn: true,
        cancelLabel: t('common:btn_close'),
        onConfirm: null,
      });
    }
  };

  const handleConfirmDelete = () => {
    actions.deleteSeries(
      id,
      // handleError
    );
  };

  const onShowMenu = () => {
    Keyboard.dismiss();
    showModal({
      isOpen: true,
      ContentComponent: (
        <MenuContent
          data={series}
          isActor={isActor}
          contentType={PostType.SERIES}
          handleConfirmDeleteSeries={handleConfirmDelete}
        />
      ),
    });
  };

  return (
    <ContentHeader
      actor={actor}
      disabled={disabled}
      audience={audience}
      onPressHeader={onPressHeader}
      onPressMenu={onShowMenu}
    />
  );
};

export default SeriesHeader;
