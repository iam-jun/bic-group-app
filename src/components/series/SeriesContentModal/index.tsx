import React, { useEffect } from 'react';
import { ExtendedTheme, useTheme } from '@react-navigation/native';
import { View, StyleSheet, FlatList } from 'react-native';
import { useRootNavigation } from '~/hooks/navigation';
import spacing from '~/theme/spacing';
import ViewSpacing from '~/beinComponents/ViewSpacing';
import SeriesItemModal from './SeriesItemModal';
import Image from '~/beinComponents/Image';
import images from '~/resources/images';
import Text from '~/baseComponents/Text';
import seriesStack from '~/router/navigator/MainStack/stacks/series/stack';
import useSeriesContentModalStore from './store';
import { IGetSeries } from '~/interfaces/ISeries';
import LoadingIndicator from '~/beinComponents/LoadingIndicator';
import useModalStore from '~/store/modal';

interface ISeriesContentModalProps {
    id: string;
}

const SeriesContentModal: React.FC<ISeriesContentModalProps> = ({ id }) => {
  const { rootNavigation } = useRootNavigation();
  const theme: ExtendedTheme = useTheme();
  const { colors } = theme;
  const styles = createStyle(theme);

  const modalActions = useModalStore((state) => state.actions);
  const { data, loading } = useSeriesContentModalStore((state) => state.series);
  const actions = useSeriesContentModalStore((state) => state.actions);
  const reset = useSeriesContentModalStore((state) => state.reset);

  useEffect(() => {
    getData();

    return reset;
  }, []);

  const getData = () => {
    const payload: IGetSeries = {
      itemIds: [id],
    };
    actions.getSeriesByItems(payload);
  };

  const onLoadMore = () => {
    getData();
  };

  const onPressClose = () => {
    modalActions.hideModal();
  };

  const onPressItem = (item: any) => {
    onPressClose();
    rootNavigation.navigate(seriesStack.seriesDetail, { seriesId: item?.id });
  };

  const renderHeaderComponent = () => <ViewSpacing height={spacing.margin.base} />;

  const renderItem = ({ item }: any) => (
    <SeriesItemModal
      data={item}
      onPressItem={() => onPressItem(item)}
    />
  );

  const renderEmptyComponent = () => {
    if (loading) return <LoadingIndicator style={styles.loading} />;

    return (
      <View style={styles.boxEmpty} testID="series_content_modal.box_empty">
        <Image
          resizeMode="contain"
          source={images.img_empty_box}
          style={styles.imgEmpty}
        />
        <Text.BodyS color={colors.neutral40} useI18n>
          common:text_no_series_modal
        </Text.BodyS>
      </View>
    );
  };

  const keyExtractor = (item, index) => `series_${item?.id || index}`;

  return (
    <View style={styles.container}>
      <FlatList
        showsVerticalScrollIndicator={false}
        data={data}
        renderItem={renderItem}
        keyExtractor={keyExtractor}
        ListEmptyComponent={renderEmptyComponent}
        ListHeaderComponent={renderHeaderComponent}
        onEndReached={onLoadMore}
        onEndReachedThreshold={0.1}
      />
    </View>
  );
};

const createStyle = (theme: ExtendedTheme) => {
  const { colors } = theme;

  return StyleSheet.create({
    container: {
      flex: 1,
      borderTopWidth: 1,
      borderTopColor: colors.neutral5,
      paddingHorizontal: 0,
      paddingBottom: 0,
    },
    boxEmpty: {
      alignItems: 'center',
      marginTop: spacing.margin.base,
      paddingVertical: 32,
    },
    imgEmpty: {
      width: 100,
      aspectRatio: 1,
      marginBottom: spacing.margin.large,
    },
    loading: {
      marginTop: spacing.margin.large,
    },
  });
};

export default SeriesContentModal;
