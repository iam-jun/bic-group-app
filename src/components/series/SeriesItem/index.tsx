import { ExtendedTheme, useTheme } from '@react-navigation/native';
import React, { FC } from 'react';
import { View, StyleSheet } from 'react-native';
import DeletedItem from '../DeletedItem';
import SeriesContent from '../SeriesContent';
import SeriesHeader from '../SeriesHeader';

type SeriesItemProps = {
    data: any;
}

const SeriesItem: FC<SeriesItemProps> = ({ data: series }) => {
  const theme = useTheme();
  const styles = createStyle(theme);

  const { deleted = false } = series || {};

  if (deleted) {
    return <DeletedItem />;
  }

  return (
    <View style={styles.container}>
      <SeriesHeader
        series={series}
        disabled={false}
      />
      <SeriesContent series={series} />
    </View>
  );
};

const createStyle = (theme: ExtendedTheme) => {
  const { colors, elevations } = theme;
  return StyleSheet.create({
    container: {
      backgroundColor: colors.white,
      ...elevations.e2,
    },
  });
};

export default SeriesItem;
