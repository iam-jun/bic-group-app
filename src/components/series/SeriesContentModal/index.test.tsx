import * as React from 'react';
import { fireEvent, renderWithRedux } from '~/test/testUtils';
import SeriesContentModal from './index';
import * as navigationHook from '~/hooks/navigation';
import seriesStack from '~/router/navigator/MainStack/stacks/series/stack';
import { mockSeries } from '~/test/mock_data/series';
import { mockArticle } from '~/test/mock_data/article';
import useSeriesContentModalStore, { ISeriesContentModalState } from './store';

describe('SeriesContentModal component', () => {
  it('should render correctly', () => {
    const rendered = renderWithRedux(
      <SeriesContentModal id={mockArticle.id} />,
    ).toJSON();

    expect(rendered).toMatchSnapshot();
  });

  it('should render empty component', () => {
    useSeriesContentModalStore.setState((state: ISeriesContentModalState) => {
      state.series.data = [];
      state.series.loading = false;
      return state;
    });

    const rendered = renderWithRedux(
      <SeriesContentModal id={mockArticle.id} />,
    );

    const emptyComponent = rendered.queryByTestId('series_content_modal.box_empty');

    expect(emptyComponent).toBeDefined();
  });

  it('should not render empty component when loading', () => {
    useSeriesContentModalStore.setState((state: ISeriesContentModalState) => {
      state.series.data = [mockSeries];
      state.series.loading = true;
      return state;
    });

    const rendered = renderWithRedux(
      <SeriesContentModal id={mockArticle.id} />,
    );

    const emptyComponent = rendered.queryByTestId('series_content_modal.box_empty');

    expect(emptyComponent).toBeNull();
  });

  it('should render navigate to SeriesDetail', () => {
    const navigate = jest.fn();
    const rootNavigation = { navigate };

    jest.spyOn(navigationHook, 'useRootNavigation').mockImplementation(() => ({ rootNavigation } as any));

    useSeriesContentModalStore.setState((state: ISeriesContentModalState) => {
      state.series.data = [mockSeries];
      state.series.loading = false;
      return state;
    });

    const rendered = renderWithRedux(
      <SeriesContentModal id={mockArticle.id} />,
    );

    const btnSeries = rendered.getByTestId('series_item_modal.button');

    expect(btnSeries).toBeDefined();
    fireEvent.press(btnSeries);
    expect(navigate).toBeCalledWith(seriesStack.seriesDetail, { seriesId: mockSeries.id });
  });
});
