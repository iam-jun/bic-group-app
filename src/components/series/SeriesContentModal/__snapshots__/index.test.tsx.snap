// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SeriesContentModal component should render correctly 1`] = `
<View
  style={
    Object {
      "borderTopColor": "#E3E6F2",
      "borderTopWidth": 1,
      "flex": 1,
      "paddingBottom": 0,
      "paddingHorizontal": 0,
    }
  }
>
  <RCTScrollView
    ListEmptyComponent={[Function]}
    ListHeaderComponent={[Function]}
    data={Array []}
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onEndReachedThreshold={0.1}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    showsVerticalScrollIndicator={false}
    stickyHeaderIndices={Array []}
    viewabilityConfigCallbackPairs={Array []}
  >
    <View>
      <View
        onLayout={[Function]}
      >
        <View
          style={
            Object {
              "height": 12,
              "paddingHorizontal": undefined,
              "paddingVertical": undefined,
              "width": undefined,
            }
          }
        />
      </View>
      <ActivityIndicator
        color="#A4A9B7"
        style={
          Object {
            "marginTop": 16,
          }
        }
        testID="loading_indicator"
      />
    </View>
  </RCTScrollView>
</View>
`;
